### ISSUE TITLE
Make metering interaction less agressive

### ISSUE THREAD

**Issue by janezpodhostnik** [onflow/flow-go#3782]
## Some Context

1. Interaction metering is done in the FVM to ensure that scripts and transactions don't read or write excessively to storage.
  interaction is measured in bytes:
  
    - every byte read from storage. Reading the same register multiple times doesn't count multiple times as it has already been cached the first time, and does not need to be fetched from storage the next time.
    - every byte written to storage. When writing to the same register multiple times during a transaction only the last write counts towards interactions, since only that is going to be persisted to storage.
    - current interaction limit is 20Mb

2. Direct contract function invocation is when the FMV directly invokes a smart contract function. These are predefined and the relevant ones are: 
- get balance, called when calling `.balance` in cadence
- get available balance, called when calling `.availableBalance` in cadence
- create account, called when calling `AuthAccount(...)` in cadence

## Problem Definition

Interaction metering is aggressive when it comes to direct contract function invocation.

When running a script (or tx) all register reads go through the scripts `Meter` is tallying up the reads. if register `a` is read twice the `Meter` is responsible to keep track and ignore the second read. 

Once a direct contract function invocation is called another (child)   `Meter` is created with the same logic but is independent. Once the direct contract function invocation is done the child meter is merged into the parent. 

This merge is the problematic bit. Currently the interaction used is simply added to the parents interaction used. This does not take into account that if the child `Meter` counted a read for register `b` that the parent `Meter` already counted the total interaction used of the child and parent actually has an overlap and cannot be summed up.

## Proposed Solution

When merging meter interaction limits for a register from meter B to meter A
- `RA[k]`  is the **R**ead of register with **k**ey in meter **A**
- `WB[k]`  is the **W**rite of register with **k**ey in meter **B**
The following rules apply:

1. `RB[k] != nil && WB[k] == nil`:
    1. `RA[k] ==nil && WA[k] == nil` -> do: `RA[k] = RB[k]`
    2. `RA[k] != nil && WA[k] == nil` -> `RA[k]` must be equal to `RB[k]` as B is reading the same register as A did. Nothing to do.
    3. `RA[k] == nil && WA[k] != nil`-> when register k is read in B the value is taken from the changed value that A metered as a write so no storage read happened in B, `RB[k]` must be equal to `WA[k]`. Nothing to do.
    4. `RA[k] != nil && WA[k] != nil` -> similar to 1.3. no storage read happened in B as the changed register was read from A.  `WA[k]` must be equal to `RB[k]`. Nothing to do
2.  `RB[k] == nil && WB[k] != nil`:
    1. `RA[k] ==nil && WA[k] == nil` -> do: `WA[k] = WB[k]`
    2. `RA[k] != nil && WA[k] == nil` -> do: `WA[k] = WB[k]`
    3. `RA[k] == nil && WA[k] != nil` -> the write in B is latest so that one should be used. do: `WA[k] = WB[k]`
    4. `RA[k] != nil && WA[k] != nil` -> the write in B is latest so that one should be used. do: `WA[k] = WB[k]`
3. `RB[k] != nil && WB[k] != nil`:
    1. `RA[k] ==nil && WA[k] == nil` -> do: `WA[k] = WB[k]` and `RA[k] = RB[k]`
    2. `RA[k] != nil && WA[k] == nil` -> `RA[k]` must be equal to `RB[k]` as B is reading the same register as A did. do: `WA[k] = WB[k]`
    3. `RA[k] == nil && WA[k] != nil` -> B read should be equal to A write, because the value of the register was taken from the changed value in A, `RB[k]` must be equal to `WA[k]`. do:  `WA[k] = WB[k]`
    4 `RA[k] != nil && WA[k] != nil` -> similar to 3.3. do:  `WA[k] = WB[k]`
	
The total interaction used should be computed from the read list and the write list, and not kept as a separate number.
